src/zjs_spi.c

```diff
@@ -0,0 +1,322 @@
+    }
+
+    size_t len;
+    jerry_value_t buffer;
+    zjs_buffer_t *tx_buf;
+    zjs_buffer_t *rx_buf;
+    jerry_value_t rx_buf_obj = jerry_create_null();
+    jerry_value_t tx_buf_obj;
+
+    if (spi_slave_select(handle->spi_device, jerry_get_number_value(argv[0]))) {
+        return zjs_error("SPI slave select failed\n");
+    }
+
+    // If only a slave is given, this must be a single read or its invalid
+    if (argc == 1) {
+        if (handle->topology != ZJS_TOPOLOGY_READ) {
+            return ZJS_STD_ERROR(RangeError, "Missing transmit buffer");
+        }
+    }
+
+    jerry_size_t dir_len = 13;
+    char dir_str[dir_len];
+    // Set the direction default based on the topology.
+    enum spi_topology dir_arg = handle->topology;
+
+    // If we have a 'direction' arg, get it and validate
+    if (argc >= 3) {
+        zjs_copy_jstring(argv[2], dir_str, &dir_len);
+
+        if (strncmp(dir_str, "read-write", 11) == 0)
+            dir_arg = ZJS_TOPOLOGY_FULL_DUPLEX;
+        else if (strncmp(dir_str, "read", 5) == 0)
+            dir_arg = ZJS_TOPOLOGY_READ;
+        else if (strncmp(dir_str, "write", 6) == 0)
+            dir_arg = ZJS_TOPOLOGY_WRITE;
+        else
+            return ZJS_STD_ERROR(Error, "Invalid direction");
+
+        // If topology conflicts with direction given
+        if ((handle->topology == ZJS_TOPOLOGY_WRITE &&
+            dir_arg != ZJS_TOPOLOGY_WRITE) ||
+            (handle->topology == ZJS_TOPOLOGY_READ &&
+            dir_arg != ZJS_TOPOLOGY_READ)) {
+            return ZJS_STD_ERROR(NotSupportedError, "Direction conflicts with topology");
+        }
+        // If reading only, the 2nd arg should be NULL
+        if (dir_arg == ZJS_TOPOLOGY_READ && !jerry_value_is_null(argv[1])) {
+            return ZJS_STD_ERROR(NotSupportedError, "Buffer should be null when direction is read");
+        }
+    }
+
+    // If we need to write a buffer
+    if (dir_arg != ZJS_TOPOLOGY_READ) {
+        buffer = argv[1];
+        // Figure out if the buffer is an array or string, handle accordingly
+        if (jerry_value_is_array(argv[1])) {
+            len = jerry_get_array_length(buffer);
+            tx_buf_obj = zjs_buffer_create(len, &tx_buf);
+            if (tx_buf) {
+                for (int i = 0; i < len; i++) {
+                    ZVAL item = jerry_get_property_by_index(buffer, i);
+                    if (jerry_value_is_number(item)) {
+                        tx_buf->buffer[i] = (uint8_t)jerry_get_number_value(item);
+                    }
+                    else {
+                        ERR_PRINT("non-numeric value in array, treating as 0\n");
+                        tx_buf->buffer[i] = 0;
+                    }
+                }
+            }
+            else {
+                 return ZJS_STD_ERROR(Error, "Failed to get TX buffer");
+            }
+        }
+        else {
+            tx_buf_obj = zjs_buffer_create(jerry_get_string_size(buffer), &tx_buf);
```

Comment:
Sorry wasn't clear where the null check was missing.  Ill check argv[1] for being null if a write is going to occur.